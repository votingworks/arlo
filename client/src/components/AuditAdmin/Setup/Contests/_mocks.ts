export const contestsInputMocks = {
  inputs: [
    { key: 'Contest Name', value: 'Contest Name' },
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
    { key: 'Total Ballot Cards Cast for Contest', value: '30' },
  ],
  errorInputs: [
    { key: 'Contest Name', value: '', error: 'Required' },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '',
      error:
        'Must be greater than or equal to the sum of votes for each candidate/choice',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

export default contestsInputMocks
