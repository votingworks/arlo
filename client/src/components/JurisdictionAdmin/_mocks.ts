import { IBatches } from './useBatchResults'
import { IRound } from '../AuditAdmin/useRoundsAuditAdmin'
import { FileProcessingStatus } from '../useCSV'
import {
  IFullHandTallyBatchResults,
  IFullHandTallyBatchResult,
} from './useFullHandTallyResults'
import { mocksOfType } from '../testUtilities'

export interface INullResultValues {
  [contestId: string]: {
    [choiceId: string]: null | string | number
  }
}

export const roundMocks = mocksOfType<IRound>()({
  incomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: false,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  complete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: '2020-09-14T17:35:19.482Z',
    needsFullHandTally: false,
    isAuditComplete: true,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  fullHandTallyIncomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: true,
    isFullHandTally: true,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
})

export const resultsMocks = mocksOfType<INullResultValues>()({
  emptyInitial: {
    'contest-id-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'contest-id-2': {
      'choice-id-3': null,
      'choice-id-4': null,
    },
  },
  complete: {
    'contest-id-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'contest-id-2': {
      'choice-id-3': 1,
      'choice-id-4': 2,
    },
  },
})

export const batchResultsMocks = mocksOfType<INullResultValues>()({
  empty: {
    'batch-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-2': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-3': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
  },
  complete: {
    'batch-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-2': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-3': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
  },
})

export const batchesMocks = mocksOfType<IBatches>()({
  emptyInitial: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
    ],
    resultsFinalizedAt: null,
  },
  complete: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 1,
              'choice-id-2': 2,
            },
          },
        ],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 0,
              'choice-id-2': 10,
            },
          },
        ],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 2000,
              'choice-id-2': 20,
            },
          },
        ],
      },
    ],
    resultsFinalizedAt: '2020-09-14T17:35:19.482Z',
  },
})

export const fullHandTallyBatchResultMock = mocksOfType<
  IFullHandTallyBatchResults
>()({
  empty: {
    finalizedAt: '',
    results: [],
  },
  complete: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  completeWithMultipleBatch: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 15,
        },
      },
      {
        batchName: 'Batch2',
        batchType: 'Provisional',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 5,
        },
      },
    ],
  },
  updated: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  finalized: {
    finalizedAt: '2021-04-13T15:01:00.383031+00:00',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
})

export const fullHandTallyBatchResultsMock = mocksOfType<
  IFullHandTallyBatchResult
>()({
  empty: {
    batchName: '',
    batchType: '',
    choiceResults: {},
  },
  complete: {
    batchName: 'Batch1',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
  updated: {
    batchName: 'Batch12',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
})
