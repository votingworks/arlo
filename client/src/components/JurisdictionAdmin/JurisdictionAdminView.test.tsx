import React from 'react'
import { screen, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ToastContainer } from 'react-toastify'
import { useParams } from 'react-router-dom'
import AuthDataProvider, { useAuthDataContext } from '../UserContext'
import JurisdictionAdminView from './JurisdictionAdminView'
import { renderWithRouter, withMockFetch, serverError } from '../testUtilities'
import {
  auditSettings,
  manifestMocks,
  talliesMocks,
  manifestFile,
  talliesFile,
  cvrsMocks,
  cvrsFile,
  roundMocks,
} from '../AuditAdmin/useSetupMenuItems/_mocks'
import { jaApiCalls } from '../_mocks'

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useRouteMatch: jest.fn(),
  useParams: jest.fn(),
}))
const paramsMock = useParams as jest.Mock

jest.mock('axios')

describe('JA setup', () => {
  // JurisdictionAdminView will only be rendered once the user is logged in, so
  // we simulate that.
  const JurisdictionAdminViewWithAuth: React.FC = () => {
    const auth = useAuthDataContext()
    return auth ? <JurisdictionAdminView /> : null
  }

  const renderView = () =>
    renderWithRouter(
      <AuthDataProvider>
        <JurisdictionAdminViewWithAuth />
        <ToastContainer />
      </AuthDataProvider>,
      {
        route: '/election/1/jurisdiction/jurisdiction-id-1/setup',
      }
    )

  beforeEach(() => {
    paramsMock.mockReturnValue({
      electionId: '1',
      jurisdictionId: 'jurisdiction-id-1',
      view: 'setup',
    })
  })

  it('renders initial state', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { container } = renderView()
      await screen.findByText('Audit Source Data')
      expect(container).toMatchSnapshot()
    })
  })

  it('submits ballot manifest and deletes it', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.all),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.deleteManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      const uploadButton = screen.getByRole('button', { name: 'Upload File' })
      userEvent.click(uploadButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(screen.getByLabelText('Select a file...'), manifestFile)
      userEvent.click(uploadButton)
      await screen.findByText('Uploaded at 6/8/2020, 9:39:14 PM.')

      // We test delete after submit so that we can check that the input is
      // cleared of the originally submitted file
      const deleteButton = await screen.findByRole('button', {
        name: 'Delete File',
      })
      userEvent.click(deleteButton)
      await screen.findByRole('button', { name: 'Upload File' })
      screen.getByLabelText('Select a file...')
    })
  })

  it('submits batch tallies', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
      jaApiCalls.putTallies,
      jaApiCalls.getBatchTalliesFile(talliesMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      const talliesInput = screen.getByLabelText('Select a file...')
      const talliesButton = screen.getByRole('button', { name: 'Upload File' })

      userEvent.click(talliesButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(talliesInput, talliesFile)
      userEvent.click(talliesButton)
      await screen.findByText('Uploaded at 7/8/2020, 9:39:14 PM.')
    })
  })

  it('toasts error when uploading batch tallies', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
      serverError('putTallies', jaApiCalls.putTallies),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      userEvent.upload(screen.getByLabelText('Select a file...'), talliesFile)
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      const toast = await screen.findByRole('alert')
      expect(toast).toHaveTextContent('something went wrong: putTallies')
    })
  })

  it('displays errors on invalid batch tallies upload', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getBatchTalliesFile(talliesMocks.processed),
      jaApiCalls.putTallies,
      jaApiCalls.getBatchTalliesFile(talliesMocks.errored),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      expect(screen.getAllByText(/Uploaded/)).toHaveLength(2)

      // Replace & upload errored batch tallies
      userEvent.click(
        screen.getAllByRole('button', {
          name: 'Replace File',
        })[1]
      )
      userEvent.upload(
        await screen.findByLabelText('Select a file...'),
        talliesFile
      )
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText('Invalid CSV')
    })
  })

  it('displays errors after reprocessing batch tallies on replacing manifest', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.getBatchTalliesFile(talliesMocks.processed),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getBatchTalliesFile(talliesMocks.errored),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      // Upload a new manifest
      userEvent.click(
        screen.getAllByRole('button', { name: 'Replace File' })[0]
      )
      userEvent.upload(
        await screen.findByLabelText('Select a file...'),
        manifestFile
      )
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText(/Uploaded/)
      await screen.findByText('Invalid CSV')
    })
  })

  it('submits CVRs', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.empty),
      jaApiCalls.putCVRs,
      jaApiCalls.getCVRSfile(cvrsMocks.processing),
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      const fileTypeSelect = screen.getByLabelText(/CVR File Type:/)
      within(fileTypeSelect).getByRole('option', {
        name: 'Dominion',
        selected: true,
      })
      userEvent.selectOptions(
        fileTypeSelect,
        screen.getByRole('option', { name: 'ClearBallot' })
      )

      const cvrsInput = screen.getByLabelText('Select a file...')
      const cvrsButton = screen.getByRole('button', { name: 'Upload File' })

      userEvent.click(cvrsButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(cvrsInput, cvrsFile)
      userEvent.click(cvrsButton)

      await screen.findByText('Uploading...')
      expect(fileTypeSelect).toBeDisabled()

      await screen.findByText('Processing...')
      expect(fileTypeSelect).toBeDisabled()

      await screen.findByText('Uploaded at 11/18/2020, 9:39:14 PM.')
      expect(fileTypeSelect).toBeDisabled()
      within(fileTypeSelect).getByRole('option', {
        name: 'ClearBallot',
        selected: true,
      })
    })
  })

  it('after deleting CVRs, keeps last selected CVR file type ', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
      jaApiCalls.deleteCVRs,
      jaApiCalls.getCVRSfile(cvrsMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Uploaded at 6/8/2020, 9:39:14 PM.')
      screen.getByRole('option', { name: 'ClearBallot', selected: true })
      userEvent.click(screen.getAllByRole('button', { name: 'Delete File' })[1])
      await screen.findByText('Select a file...')
      screen.getByRole('option', { name: 'ClearBallot', selected: true })
    })
  })

  it('displays errors on invalid CVRs upload', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
      jaApiCalls.putCVRs,
      jaApiCalls.getCVRSfile(cvrsMocks.errored),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      expect(screen.getAllByText(/Uploaded/)).toHaveLength(2)

      // Replace & upload errored CVRs
      userEvent.click(
        screen.getAllByRole('button', {
          name: 'Replace File',
        })[1]
      )
      userEvent.upload(
        await screen.findByLabelText('Select a file...'),
        cvrsFile
      )
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText('Invalid CSV')
    })
  })

  it('allows multiple CVR files to be uploaded for ES&S', async () => {
    const cvrsFormData: FormData = new FormData()
    const cvrsFile1 = new File(['test cvr data'], 'cvrs1.csv', {
      type: 'text/csv',
    })
    const cvrsFile2 = new File(['test cvr data'], 'cvrs2.csv', {
      type: 'text/csv',
    })
    // Make the combined CVR files large enough to trigger an "Uploading..." progress bar
    Object.defineProperty(cvrsFile1, 'size', { value: 500 * 1000 })
    Object.defineProperty(cvrsFile2, 'size', { value: 500 * 1000 })
    cvrsFormData.append('cvrs', cvrsFile1, cvrsFile1.name)
    cvrsFormData.append('cvrs', cvrsFile2, cvrsFile2.name)
    cvrsFormData.append('cvrFileType', 'ESS')

    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.empty),
      {
        ...jaApiCalls.putCVRs,
        options: { ...jaApiCalls.putCVRs.options, body: cvrsFormData },
      },
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      userEvent.selectOptions(
        screen.getByLabelText(/CVR File Type:/),
        screen.getByRole('option', { name: 'ES&S' })
      )

      const fileSelect = screen.getByLabelText('Select files...')
      userEvent.upload(fileSelect, [cvrsFile1, cvrsFile2])
      await screen.findByLabelText('2 files selected')
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))

      await screen.findByText('Uploading...')
      await screen.findByText('Uploaded at 11/18/2020, 9:39:14 PM.')
    })
  })

  it('allows CVRs ZIP file to be uploaded for Hart', async () => {
    const cvrsFormData: FormData = new FormData()
    const cvrsZip = new File(['test cvr data'], 'cvrs.zip', {
      type: 'application/zip',
    })
    cvrsFormData.append('cvrs', cvrsZip, cvrsZip.name)
    cvrsFormData.append('cvrFileType', 'HART')

    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.empty),
      {
        ...jaApiCalls.putCVRs,
        options: { ...jaApiCalls.putCVRs.options, body: cvrsFormData },
      },
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      userEvent.selectOptions(
        screen.getByLabelText(/CVR File Type:/),
        screen.getByRole('option', { name: 'Hart' })
      )

      const fileSelect = screen.getByLabelText('Select files...')
      userEvent.upload(fileSelect, [cvrsZip])
      await screen.findByLabelText('cvrs.zip')
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText('Uploaded at 11/18/2020, 9:39:14 PM.')
    })
  })

  it('allows CVRs ZIP file and scanned ballot information CSV to be uploaded for Hart', async () => {
    const cvrsFormData: FormData = new FormData()
    const cvrsZip = new File(['test cvr data'], 'cvrs.zip', {
      type: 'application/zip',
    })
    const scannedBallotInformationCsv = new File(
      ['test cvr data'],
      'scanned-ballot-information.csv',
      {
        type: 'text/csv',
      }
    )
    cvrsFormData.append('cvrs', cvrsZip, cvrsZip.name)
    cvrsFormData.append(
      'cvrs',
      scannedBallotInformationCsv,
      scannedBallotInformationCsv.name
    )
    cvrsFormData.append('cvrFileType', 'HART')

    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.empty),
      {
        ...jaApiCalls.putCVRs,
        options: { ...jaApiCalls.putCVRs.options, body: cvrsFormData },
      },
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      userEvent.selectOptions(
        screen.getByLabelText(/CVR File Type:/),
        screen.getByRole('option', { name: 'Hart' })
      )
      await screen.findByText(
        /For Hart, you can also provide an optional scanned ballot information CSV. If provided, the unique identifiers in the CSV will be used as imprinted IDs./
      )

      const fileSelect = screen.getByLabelText('Select files...')
      userEvent.upload(fileSelect, [cvrsZip, scannedBallotInformationCsv])
      await screen.findByLabelText('2 files selected')
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText('Uploaded at 11/18/2020, 9:39:14 PM.')
    })
  })

  it('displays errors after reprocessing CVRs on replacing manifest', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.ballotComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.getCVRSfile(cvrsMocks.processed),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getCVRSfile(cvrsMocks.errored),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      // Upload a new manifest
      userEvent.click(
        screen.getAllByRole('button', { name: 'Replace File' })[0]
      )
      userEvent.upload(
        await screen.findByLabelText('Select a file...'),
        manifestFile
      )
      userEvent.click(screen.getByRole('button', { name: 'Upload File' }))
      await screen.findByText(/Uploaded/)
      await screen.findByText('Invalid CSV')
    })
  })

  it('shows error with incorrect file', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.errored),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')
      const [manifestInput, talliesInput] = screen.getAllByLabelText(
        'Select a file...'
      )
      const [manifestButton, talliesButton] = screen.getAllByRole('button', {
        name: 'Upload File',
      })

      expect(talliesInput).toBeDisabled()
      expect(talliesButton).toBeDisabled()

      userEvent.click(manifestButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(manifestInput, manifestFile)
      userEvent.click(manifestButton)
      await screen.findByText('Invalid CSV')
    })
  })

  it('replaces manifest file', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.batchComparisonAll),
      jaApiCalls.getRounds([]),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      jaApiCalls.getBatchTalliesFile(talliesMocks.empty),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByText('Audit Source Data')

      const replaceButton = await screen.findByText('Replace File')
      userEvent.click(replaceButton)
      const inputFiles = await screen.findAllByText('Select a file...')
      expect(inputFiles).toHaveLength(2)

      const [manifestInput] = screen.getAllByLabelText('Select a file...')
      const [manifestButton] = screen.getAllByRole('button', {
        name: 'Upload File',
      })

      userEvent.click(manifestButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(manifestInput, manifestFile)
      userEvent.click(manifestButton)
      await screen.findByText('Current file:')
    })
  })

  it('stays on the file upload screen when sample is being drawn', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.all),
      jaApiCalls.getRounds(roundMocks.drawSampleInProgress),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByRole('heading', { name: 'Audit Source Data' })
      screen.getByText('The audit has not started.')
    })
  })

  it('stays on the file upload screen when drawing sample errors', async () => {
    const expectedCalls = [
      jaApiCalls.getUser,
      jaApiCalls.getSettings(auditSettings.all),
      jaApiCalls.getRounds(roundMocks.drawSampleErrored),
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
    ]
    await withMockFetch(expectedCalls, async () => {
      renderView()
      await screen.findByRole('heading', { name: 'Audit Source Data' })
      screen.getByText('The audit has not started.')
    })
  })
})
