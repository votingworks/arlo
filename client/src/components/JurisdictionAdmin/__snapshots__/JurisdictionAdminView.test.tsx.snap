// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JA setup renders initial state 1`] = `
<div>
  <div
    class="sc-bdVaJa bKocXA"
  >
    <div
      class="bp3-callout sc-bxivhb dweKfW"
    >
      <div
        class="sc-bwzfXH dkQuHa"
      >
        <div
          class="text"
        >
          <h3
            class="bp3-heading sc-htpNat hodhlw"
          >
            Test Audit
          </h3>
          <h4
            class="bp3-heading"
          >
            The audit has not started.
          </h4>
          <p>
            0/2 files successfully uploaded.
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-bwzfXH sc-brqgnP hXccOg"
    >
      <h2
        class="bp3-heading sc-ifAKCX cmZyyF"
      >
        Audit Source Data
      </h2>
      <form>
        <div
          class="sc-EHOje bjGJmd"
        >
          <h2
            class="bp3-heading sc-ifAKCX cmZyyF"
          />
          <div
            class="sc-bZQynM bGsEQi"
          >
            <h4
              class="bp3-heading"
            >
              Ballot Manifest
            </h4>
            <div
              class="sc-gzVnrw kvvweo"
            >
              Click "Browse" to choose the appropriate Ballot
                Manifest file from your computer. This file should be a
                comma-separated list of all the ballot boxes/containers used
                to store ballots for this particular election, plus a count of
                how many ballot cards (individual pieces of paper) are stored
                in each container.
              <br />
              <br />
              <a
                href="/sample_ballot_manifest.csv"
                rel="noopener noreferrer"
                target="_blank"
              >
                (Click here to view a sample file in the correct format.)
              </a>
            </div>
          </div>
          <div
            class="sc-bZQynM bGsEQi"
          >
            <label
              class="bp3-file-input"
            >
              <input
                accept=".csv"
                name="csv"
                type="file"
              />
              <span
                class="bp3-file-upload-input"
              >
                Select a file...
              </span>
            </label>
            <div
              style="display: flex; margin-top: 15px; margin-bottom: 10px; align-items: center; width: 300px;"
            />
            <button
              class="bp3-button bp3-intent-primary sc-gZMcBi eZVGYn"
              type="submit"
            >
              <span
                class="bp3-button-text"
              >
                Upload File
              </span>
            </button>
          </div>
        </div>
      </form>
      <form>
        <div
          class="sc-EHOje bjGJmd"
        >
          <h2
            class="bp3-heading sc-ifAKCX cmZyyF"
          />
          <div
            class="sc-bZQynM bGsEQi"
          >
            <h4
              class="bp3-heading"
            >
              Candidate Totals by Batch
            </h4>
            <div
              class="sc-gzVnrw kvvweo"
            >
              Click "Browse" to choose the appropriate Candidate Totals by Batch file from your computer. This file should be a comma-separated list of all the ballot boxes/containers used to store ballots for this particular election, plus a count of how many votes were counted for each candidate in each of those containers.
              <br />
              <br />
              <a
                href="/sample_candidate_totals_by_batch.csv"
                rel="noopener noreferrer"
                target="_blank"
              >
                (Click here to view a sample file in the correct format.)
              </a>
            </div>
          </div>
          <div
            class="sc-bZQynM bGsEQi"
          >
            <label
              class="bp3-file-input bp3-disabled"
            >
              <input
                accept=".csv"
                disabled=""
                name="csv"
                type="file"
              />
              <span
                class="bp3-file-upload-input"
              >
                Select a file...
              </span>
            </label>
            <div
              style="display: flex; margin-top: 15px; margin-bottom: 10px; align-items: center; width: 300px;"
            />
            <button
              class="bp3-button bp3-disabled bp3-intent-primary sc-gZMcBi eZVGYn"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <span
                class="bp3-button-text"
              >
                Upload File
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
